ql/cashflows/cashflows.cpp

```diff
@@ -625,9 +625,18 @@ namespace QuantLib {
                     }
                     refEndDate = couponDate;
                 }
-
-                t += dc.yearFraction(lastDate, couponDate,
-                                     refStartDate, refEndDate);
+                
+                if (coupon && lastDate!=refStartDate){
```

Comment:
I would compare `lastDate` to the start of the coupon rather than `refStartDate` (they might be different in case of a short or long coupon).